error[E0308]: mismatched types yaml Arg::short

<!--
Please use the following template to assist with creating an issue and to ensure a speedy resolution. If an area is not applicable, feel free to delete the area or mark with `N/A`
-->

### Rust Version

* Use the output of `rustc -V`

rustc 1.31.0-nightly (1cf82fd9c 2018-10-30)

### Affected Version of clap

* Can be found in Cargo.lock of your project (i.e. `grep clap Cargo.lock`)

3.0.0-alpha.1

### Bug or Feature Request Summary

94872e00a50717c3c59f39bb8a949c2116b952fc broke yaml

### Expected Behavior Summary

Yaml is parsed

### Actual Behavior Summary

Yaml is not successfully parsed

### Steps to Reproduce the issue

Enable the "yaml" feature

### Sample Code or Link to Sample Code

[Minimal Cargo.toml #1368](https://gist.github.com/andrewmiller1/9e5456f236d9ff9fc0051c8acfc7e82c)

[Minimal working Cargo.toml #1368](https://gist.github.com/andrewmiller1/375ec96892fc9a71ac35af007965faa4) (uses 683105cdd10c8a301e043a6c77531b2ff8c6a383)

Since the build fails when compiling clap, the other two files aren't relevant.

~[Minimal main.rs #1368](https://gist.github.com/andrewmiller1/c8e582f0627cccdc7d1a1676d6b2610a)~

~[Minimal clap yaml #1368](https://gist.github.com/andrewmiller1/f36a38ffd960dd20d00e2b2643591bf4)~

### Debug output

Compile clap with cargo features `"debug"` such as:

```toml
[dependencies]
clap = { version = "2", features = ["debug"] }
```

<details>
<summary> Debug Output </summary>
<pre>
<code>

andrew@pop-os:~/projects/dev-old/targets/dev (master *%)
$ build --verbose
       Fresh unicode-xid v0.1.0                                                                                                                                                                                                              
       Fresh libc v0.2.43                                                                                                                                                                                                                    
       Fresh linked-hash-map v0.5.1                                                                                                                                                                                                          
       Fresh unicode-width v0.1.5                                                                                                                                                                                                            
       Fresh bitflags v1.0.4                                                                                                                                                                                                                 
       Fresh indexmap v1.0.2                                                                                                                                                                                                                 
       Fresh strsim v0.7.0                                                                                                                                                                                                                   
       Fresh vec_map v0.8.1                                                                                                                                                                                                                  
       Fresh ansi_term v0.11.0                                                                                                                                                                                                               
       Fresh atty v0.2.11                                                                                                                                                                                                                    
       Fresh yaml-rust v0.4.2                                                                                                                                                                                                                
       Fresh textwrap v0.10.0                                                                                                                                                                                                                
       Fresh proc-macro2 v0.4.20                                                                                                                                                                                                             
       Fresh quote v0.6.9                                                                                                                                                                                                                    
       Fresh syn v0.14.9                                                                                                                                                                                                                     
       Fresh clap_derive v0.3.0 (https://github.com/clap-rs/clap_derive#2fad2c8a)                                                                                                                                                            
   Compiling clap v3.0.0-alpha.1 (https://github.com/clap-rs/clap?branch=v3-master#eaa0700e)                                                                                                                                                 
     Running `rustc --crate-name clap /home/andrew/.cargo/git/checkouts/clap-78dbe9b58f9073fe/eaa0700/src/lib.rs --color always --crate-type lib --emit=dep-info,link -C debuginfo=2 --cfg 'feature="ansi_term"' --cfg 'feature="atty"' --cfg 'feature="clap_derive"' --cfg 'feature="color"' --cfg 'feature="debug"' --cfg 'feature="default"' --cfg 'feature="derive"' --cfg 'feature="strsim"' --cfg 'feature="suggestions"' --cfg 'feature="unstable"' --cfg 'feature="vec_map"' --cfg 'feature="yaml"' --cfg 'feature="yaml-rust"' -C metadata=e1383ef14033cd3b -C extra-filename=-e1383ef14033cd3b --out-dir /home/andrew/projects/dev/target/debug/deps -L dependency=/home/andrew/projects/dev/target/debug/deps --extern ansi_term=/home/andrew/projects/dev/target/debug/deps/libansi_term-6a1b6f7dd99c1473.rlib --extern atty=/home/andrew/projects/dev/target/debug/deps/libatty-32a142521ad608c5.rlib --extern bitflags=/home/andrew/projects/dev/target/debug/deps/libbitflags-a91b077889eb443f.rlib --extern clap_derive=/home/andrew/projects/dev/target/debug/deps/libclap_derive-7de525ba7d38dace.so --extern indexmap=/home/andrew/projects/dev/target/debug/deps/libindexmap-aff436b5b93fd6ec.rlib --extern strsim=/home/andrew/projects/dev/target/debug/deps/libstrsim-df6c1dbbc0d0dbd1.rlib --extern textwrap=/home/andrew/projects/dev/target/debug/deps/libtextwrap-048d110e2204e4ab.rlib --extern unicode_width=/home/andrew/projects/dev/target/debug/deps/libunicode_width-1f1ddf8d3aaa7342.rlib --extern vec_map=/home/andrew/projects/dev/target/debug/deps/libvec_map-4bda7b99adb92f1a.rlib --extern yaml_rust=/home/andrew/projects/dev/target/debug/deps/libyaml_rust-2bb40893dd047a0a.rlib --cap-lints allow`
error[E0308]: mismatched types                                                                                                                                                                                                               
   --> /home/andrew/.cargo/git/checkouts/clap-78dbe9b58f9073fe/eaa0700/src/build/macros.rs:82:9                                                                                                                                              
    |                                                                                                                                                                                                                                        
82  | /         $v.as_str()                                                                                                                                                                                                                  
83  | |             .unwrap_or_else(|| panic!("failed to convert YAML {:?} value to a string", $v))                                                                                                                                          
    | |___________________________________________________________________________________________^ expected char, found &str                                                                                                                
    |                                                                                                                                                                                                                                        
   ::: /home/andrew/.cargo/git/checkouts/clap-78dbe9b58f9073fe/eaa0700/src/build/arg/mod.rs:166:28                                                                                                                                           
    |                                                                                                                                                                                                                                        
166 |                   "short" => yaml_to_str!(a, v, short),                                                                                                                                                                                
    |                              ------------------------- in this macro invocation                                                                                                                                                        
    |                                                                                                                                                                                                                                        
    = note: expected type `char`                                                                                                                                                                                                             
               found type `&str`                                                                                                                                                                                                             
                                                                                                                                                                                                                                             
error: aborting due to previous error                                                                                                                                                                                                        
                                                                                                                                                                                                                                             
For more information about this error, try `rustc --explain E0308`.                                                                                                                                                                          
error: Could not compile `clap`.                                                                                                                                                                                                             

Caused by:
  process didn't exit successfully: `rustc --crate-name clap /home/andrew/.cargo/git/checkouts/clap-78dbe9b58f9073fe/eaa0700/src/lib.rs --color always --crate-type lib --emit=dep-info,link -C debuginfo=2 --cfg 'feature="ansi_term"' --cfg 'feature="atty"' --cfg 'feature="clap_derive"' --cfg 'feature="color"' --cfg 'feature="debug"' --cfg 'feature="default"' --cfg 'feature="derive"' --cfg 'feature="strsim"' --cfg 'feature="suggestions"' --cfg 'feature="unstable"' --cfg 'feature="vec_map"' --cfg 'feature="yaml"' --cfg 'feature="yaml-rust"' -C metadata=e1383ef14033cd3b -C extra-filename=-e1383ef14033cd3b --out-dir /home/andrew/projects/dev/target/debug/deps -L dependency=/home/andrew/projects/dev/target/debug/deps --extern ansi_term=/home/andrew/projects/dev/target/debug/deps/libansi_term-6a1b6f7dd99c1473.rlib --extern atty=/home/andrew/projects/dev/target/debug/deps/libatty-32a142521ad608c5.rlib --extern bitflags=/home/andrew/projects/dev/target/debug/deps/libbitflags-a91b077889eb443f.rlib --extern clap_derive=/home/andrew/projects/dev/target/debug/deps/libclap_derive-7de525ba7d38dace.so --extern indexmap=/home/andrew/projects/dev/target/debug/deps/libindexmap-aff436b5b93fd6ec.rlib --extern strsim=/home/andrew/projects/dev/target/debug/deps/libstrsim-df6c1dbbc0d0dbd1.rlib --extern textwrap=/home/andrew/projects/dev/target/debug/deps/libtextwrap-048d110e2204e4ab.rlib --extern unicode_width=/home/andrew/projects/dev/target/debug/deps/libunicode_width-1f1ddf8d3aaa7342.rlib --extern vec_map=/home/andrew/projects/dev/target/debug/deps/libvec_map-4bda7b99adb92f1a.rlib --extern yaml_rust=/home/andrew/projects/dev/target/debug/deps/libyaml_rust-2bb40893dd047a0a.rlib --cap-lints allow` (exit code: 1)
andrew@pop-os:~/projects/dev-old/targets/dev (master *%)
$ 

</code>
</pre>
</details>

--- a/examples/17_yaml.rs
+++ b/examples/17_yaml.rs
@@ -13,6 +13,8 @@
 
 // Using yaml requires calling a clap macro `load_yaml!()` so we must use the '#[macro_use]'
 // directive
+// Note: If you're using clap as a dependency and don't have a feature for your users called
+// "yaml", you'll need to remove the #[cfg(feature = "yaml")] conditional compilation attribute
 #[cfg(feature = "yaml")]
 #[macro_use]
 extern crate clap;

--- a/src/build/arg/mod.rs
+++ b/src/build/arg/mod.rs
@@ -163,7 +163,7 @@ impl<'a, 'b> Arg<'a, 'b> {
 
         for (k, v) in arg_settings.iter() {
             a = match k.as_str().unwrap() {
-                "short" => yaml_to_str!(a, v, short),
+                "short" => yaml_to_char!(a, v, short),
                 "long" => yaml_to_str!(a, v, long),
                 "aliases" => yaml_vec_or_str!(v, a, alias),
                 "help" => yaml_to_str!(a, v, help),

--- a/src/build/macros.rs
+++ b/src/build/macros.rs
@@ -77,10 +77,28 @@ macro_rules! yaml_opt_str {
 }
 
 #[cfg(feature = "yaml")]
+macro_rules! yaml_char {
+    ($v:expr) => {{
+        $v.as_str()
+            .unwrap_or_else(|| panic!("failed to convert YAML {:?} value to a string", $v))
+            .chars()
+            .next()
+            .unwrap_or_else(|| panic!("Expected char"))
+    }};
+}
+
+#[cfg(feature = "yaml")]
 macro_rules! yaml_str {
     ($v:expr) => {{
         $v.as_str()
             .unwrap_or_else(|| panic!("failed to convert YAML {:?} value to a string", $v))
+    }};
+}
+
+#[cfg(feature = "yaml")]
+macro_rules! yaml_to_char {
+    ($a:ident, $v:ident, $c:ident) => {{
+        $a.$c(yaml_char!($v))
     }};
 }
 

